src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRangeHandler.cs

```diff
@@ -64,5 +74,35 @@ public SemanticTokensRangeHandler(IGlobalOptionService globalOptions)
 
             return new RoslynSemanticTokens { Data = tokensData, IsFinalized = isFinalized };
         }
+
+        private async Task<bool> IsDataFinalizedAsync(Project project, CancellationToken cancellationToken)
+        {
+            // We use a combination of IsFullyLoaded + the completed project compilation as the metric
+            // for isFinalized. It may not be completely accurate but this is only a a temporary fix until
+            // workspace/semanticTokens/refresh is implemented.
+            if (_projectIdToCompilation.TryGetValue(project.Id, out var compilationTask) && compilationTask.IsCompleted)
+            {
+                // We don't want to hang on to the compilation since this can be very expensive,
+                // but we do want to mark the compilation as being successfully retrieved.
+                if (compilationTask.Result is not null)
+                {
+                    _projectIdToCompilation[project.Id] = Task.FromResult<Compilation?>(null);
+                }
+
+                return true;
+            }
+
+            var workspaceStatusService = project.Solution.Workspace.Services.GetRequiredService<IWorkspaceStatusService>();
+            var isFullyLoaded = await workspaceStatusService.IsFullyLoadedAsync(cancellationToken).ConfigureAwait(false);
+            if (isFullyLoaded && !_projectIdToCompilation.ContainsKey(project.Id))
+            {
+                // If the project's compilation isn't yet available, kick off a task in the background to
+                // hopefully make it available faster since we'll need it later to compute tokens.
+                var newCompilationTask = project.GetCompilationAsync(cancellationToken);
+                _projectIdToCompilation.Add(project.Id, newCompilationTask);
```

Comment:
this can technically race and lead to multiple calls to `_projectIdToCompilation.Add` which would throw -
1.  2 semantic tokens requests come in for the same project (could be diff docs).
2.  They run in parallel since semantic tokens isn't mutating
3.  Running in parallel, they both hit the if `!_projectIdToCompilation.ContainsKey(project.Id)`, and they both say it is not present in the dictionary (because neither has created the task yet).
4.  Now both will try to add to the dictionary and potentially throw.

I think you'll want to guard access to the `_projectIdToCompilation` with a lock

Reply:
Done! Thanks